### ISSUE TITLE
01. First model generalized to CPJUMP1 compound plates

### ISSUE THREAD

**Issue by EchteRobert** [EchteRobert/FeatureAggregation_single_cell#1]
This first line of experiments will be designed to beat the PR baseline of profiles created with the current aggregation method (which takes the mean). All information about the compound plates used can be found here: https://github.com/jump-cellpainting/2021_Chandrasekaran_submitted.

**Comment by EchteRobert** [comment]
### Calculating the baseline
The baseline in terms of percent replicating (PR) to beat, can be calculated in a variety of ways. Replicates are considered same compound perturbations. Methods that have been tried to calculate PR metrics: 
1. according to a 2 plate train/2 plate val split used in the earliest model iterations
2. according to the 80/20 train/val split used as in the model
3. according to the default all 4 plates

Note that the train/val split methods are used to create fair comparison between data the model has and has not seen. This does introduce some problems, so in the future a complete hold-out dataset will be used for fair comparison. The second method will not be shown here, as this method is not very compatible with calculating the PR for various reasons. 

Negcons are removed from the dataset both during training and evaluation with PR.

### 2 plate training (BR00117010, BR00117011) and 2 plate validation (BR00117012, BR00117013)
PR is calculated here with nr_replicates = 3. From top to bottom: 
1. Training PR - mean aggregated profiles, no feature selection
2. Validation PR - mean aggregated profiles, no feature selection
![BENCHMARK_TVsplit_noNegcon_pertIname_nR3_PR](https://user-images.githubusercontent.com/62173977/154143797-a5ba3e01-c49b-40eb-98fa-4974af7cbc4e.png)



### default PR (as reported on the CPJUMP1 GitHub) 
PR is calculated here with nr_replicates = 6:
1. All wells PR - mean aggregated profiles, no feature selection
![pertIname_noNegcon_nR6_PR](https://user-images.githubusercontent.com/62173977/154314357-a533c002-8437-407a-90a1-961bd8a1d083.png)

**Comment by EchteRobert** [comment]
### Experiment:
The model architecture consists of 3 simple layers followed by a max pooling operation, which transforms the single-cell features into an aggregated profile in a different feature space. A projection head on top of this feature space projects the aggregated features into a different space. This final representation is then used to calculate the loss. The first model that is shown to overfit the data is trained on the 2 plate train/ 2 plate val split of the data. The loss is the SupCon loss.

### Main Takeaway:
**From this we can deduce that the loss function (inversely) correlates nicely with the PR.** 

_Loss curves_
<img width="1161" alt="TrainValLoss_overfit" src="https://user-images.githubusercontent.com/62173977/154146307-f0597fda-6cf1-43c9-8bed-6828bba15c71.png">

_The corresponding PR scores_
![TVsplit_MLP_noNegcon_pertIname_nR3_PR](https://user-images.githubusercontent.com/62173977/154146354-6295d47c-2af2-411d-8f4b-2baf780e7cf4.png)

**Comment by EchteRobert** [comment]
### Experiment
In order to get the model to generalize to the feature aggregation problem, instead of memorizing the training set, I added more data to the training. I did this in two ways:
1. I aggregated all well data, sorted it according to compound names and then split the data into 80% training and 20% validation. This means the validation set will contain compounds the training set has not seen before. 
2. Up until now `nr_cells`, the number of cells I sampled from each well, was set to 400. There are wells that contain less and more (see histogram below). I now changed that to sample multiple times from wells which contained more than 400 cells. More specifically, from each well I sampled: once if less than `nr_cells`, twice if less than 2*`nr_cells`, but more than `nr_cells`, or thrice if more than 2*`nr_cells`.

This lead to the first generalized model, which, after some tuning, showed loss curves like shown below. 

I calculated the PR both for 2 plate train/val split and for the entire dataset (all plates). I only show the PR of all plates here. I will use the model labels as shown in the legend of the loss curves.

### Main takeaways
1. Using the methods described above, results in a lower training and validation loss, as expected.
2. Seeing all compound types during training ('semiGeneralized' model) significantly increases PR of a model's profiles while the corresponding validation loss can be similar to other models, which achieve lower PR. Although expected, it is good to note that this can happen. 
3. As the model is tuned to achieve lower training and validation losses the PR seems to decrease. Unexpected result as it contradicts what was found in the previous experiment. 
The only difference I can think of is that the model is trained on positive pairs which mostly come from the same well (but different cells), while in this PR comparison we are only looking at positive pairs between different wells (but the same compound). The difference in performance between the two may then be caused by the model not learning the representations as well as hoped OR the fact that the random subsample of `nr_cells` (e.g. 400) cells is not representative enough in some cases.

![NrCellsHist](https://user-images.githubusercontent.com/62173977/154147010-00ddc902-d053-4444-ac75-55a9d14a5024.png)

_Important note here, the 'semiGeneralized model' did not have its data sorted before splitting it into train/val. That means the model has seen all compound types during training, but has not seen them from every plate:_ 
<img width="940" alt="TrainValLoss_Generalized" src="https://user-images.githubusercontent.com/62173977/154147965-f65dad92-e5d8-4dbd-8471-4a9de19d1a03.png">

_SemiGeneralized_
![AllPlates_semiGeneralized_PR](https://user-images.githubusercontent.com/62173977/154154988-49577aa0-fa8c-40f2-a6df-f22a698c3083.png)

_Generalized_
![AllPlates_Generalized_PR](https://user-images.githubusercontent.com/62173977/154155012-872fda17-10cb-4240-908d-c280571444ea.png)

_Generalized and Tuned_
![AllPlates_generalizedTuned_PR](https://user-images.githubusercontent.com/62173977/154155073-3faeec85-90cc-445a-8045-e04222ba0905.png)

**Comment by EchteRobert** [comment]
### Experiment
To test if this unexpected correlation between the loss and PR is correlated with the way the model is trained, especially how the batches are organized, I trained the same model "Generalized and Tuned" but with a batch size of 32 and 64 instead of 16, which was used above. 

### Main takeaways

- There is a strong correlation between the batch size and the PR replicating. 
- The loss is also involved, but the absolute value of the loss should be scaled to the used batch size. 
- This does not mean that the loss and PR are not correlated, but the training batches need to contain complex enough examples for the model to learn a generally good aggregation of the data. 

I will investigate how to stabilize the influence of the batch size now. 

_Loss curves of the two models_
![TrainValLoss_all](https://user-images.githubusercontent.com/62173977/154344046-cc9543ae-6d44-4725-b913-bf45386afb72.png)


_PR of the model with BS32_
![AllPlates_generalizedTuned_BS32_PR](https://user-images.githubusercontent.com/62173977/154315229-8b4813e3-1de8-4144-a33f-04a20b42400b.png)


_PR of the model with BS64_
![AllPlates_generalizedTuned_BS64_PR](https://user-images.githubusercontent.com/62173977/154344719-e2abdfe6-649a-47d0-8bc9-87d0265221a1.png)
